File: MinimumFailingPathSum.java
// Time Complexity : O(n2)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// repeated patterns liek min of children is identified

class MinimumFallingPathSum {
    public int minFallingPathSum(int[][] matrix) {
        if(matrix==null||matrix.length==0){
            return 0;
        }
        int n = matrix.length;
        for(int i=n-2; i>=0;i--){
            for(int j=0;j<n;j++){
                if(j==0){
                    matrix[i][j] = matrix[i][j] + Math.min(matrix[i+1][j], matrix[i+1][j+1]);
                } else if(j==n-1){
                    matrix[i][j] = matrix[i][j] + Math.min(matrix[i+1][j], matrix[i+1][j-1]);
                } else {
                    matrix[i][j] = matrix[i][j] + Math.min(matrix[i+1][j], Math.min(matrix[i+1][j-1], matrix[i+1][j+1]));
                }
            }
        }
        int min = Integer.MAX_VALUE;
        for(int j=0;j<n;j++){
            min = Math.min(min, matrix[0][j]);
        }
        return min;
    }

    public static void main(String[] args) {
        int [][] matrix = {{2,1,3},{6,5,4},{7,8,9}};
        MinimumFallingPathSum obj = new MinimumFallingPathSum();
        System.out.println((obj.minFallingPathSum(matrix)));
    }
}

File: minimum-failing-path-sum.py
'''
Time complexity: O(n^2)
Space complexity: O(n) where n is the number of columns in the matrix
Did this code successfully run on Leetcode : Yes(#931)
Any problem you faced while coding this : No
Approach:
- Failling def -> dp[i][j] = matrix[i][j] + min(dp[i-1][j-1],dp[i-1][j],dp[i-1][j+1])
- We can optimize the space complexity by using only 1D array dp[j] = matrix[i][j] + min(dp[j-1],dp[j],dp[j+1])
- We can use tmp_dp to avoid overridding of values
'''

from typing import List


class Solution:
    def minFallingPathSum(self, matrix: List[List[int]]) -> int:
        dp = matrix[0][:]
        for row in range(1,len(matrix)):
            tmp_dp = [0] * len(matrix[0]) # use this to avoid overridding of values
            for col in range(len(matrix[0])):
                if col-1 >=0 and col+1 <= len(matrix[0])-1:
                    min_val = min(dp[col-1],dp[col],dp[col+1])
                else:
                    if col-1 >=0: # first column
                        min_val = min(dp[col-1],dp[col])
                    else: # last column
                        min_val = min(dp[col],dp[col+1])  
                tmp_dp[col] = matrix[row][col] + min_val
            dp = tmp_dp
        return min(dp)
    

#Driver code
sol = Solution()
print(sol.minFallingPathSum([[2,1,3],[6,5,4],[7,8,9]])) # 13